### DISCUSSION TITLE
Authentication and email issues on Termux

### DISCUSSION THREAD

**Discussion by thanhdanh27600** [pocketbase/pocketbase#4030] [Q&A] [answered]
I'm trying to deploy on Termux, the authentication process (Google) throws an error message related to a TCP connection issue. Also, I'm unable to send test emails.

`Post "https://accounts.google.com/o/oauth2/token": dial tcp: lookup accounts.google.com on [::1]:53: read udp [::1]:37783->[::1]:53: read: connection refused`

![image](https://github.com/pocketbase/pocketbase/assets/47974698/02e1695e-8507-4f1c-947c-80b76ecf656e)

Has anyone encountered and resolved these issues?

Much thank. Super neat backend 🙂

**Comment by whmeitzler** [comment]
Do you mean Termux, the terminal emulator on android? That's a bold move, Cotton! I'm very curious how it goes.

Looks like a DNS lookup is failing. I'm guessing that termux is sandboxing your instance, and therefore PB is not able to hit a local DNS resolver (UDP localhost port 53). Makes sense that stock Pocketbase would only need DNS resolution for oauth.

**Reply by thanhdanh27600** [reply]
Yes, I'm going with the prebuilt Linux arm64. Everything seems smooth and fast!

**Comment by ganigeorgiev** [comment, marked as answer]
In addition to what @whmeitzler, are you using a custom Android PocketBase binary or the prebuilt Linux arm64 one?

If the latter then maybe that is the cause. From the Go documentation about [Name resolution](https://pkg.go.dev/net#hdr-Name_Resolution):

> On Unix systems, the resolver has two options for resolving names. It can use a pure Go resolver that sends DNS requests directly to the servers listed in /etc/resolv.conf, or it can use a cgo-based resolver that calls C library routines such as getaddrinfo and getnameinfo.

My guess is that your Android setup doesn't have `/etc/resolv.conf` or is located somewhere else.

You can try to set the env variable `GODEBUG=netdns=cgo` OR build PocketBase specifically for the Android device and explicitly enable CGO so that it can use the second option listed in the above documentation. Note that cross-compilation may not work in this case because the binary will be dynamically linked and it may be required to build it on the Android device itself:

1. [Install Go 1.21+](https://go.dev/doc/install)
2. Clone/download the repo
3. Navigate to `examples/base`
4. Run `GOOS=android GOARCH=arm64 CGO_ENABLED=1 go build` (_https://go.dev/doc/install/source#environment_)
4. Start the created executable by running `./base serve`.

Another option I guess could be to try to install and run a local DNS server, something like [`dnsmasq`](https://linux.die.net/man/8/dnsmasq).

**Reply by ganigeorgiev** [reply]
_See also https://github.com/golang/go/issues/10714 and https://github.com/golang/go/commit/7165c9b30e36959dcf8d0a038c803b4d2d8be445. It seems that Go has a special case when GOOS is Android and will force the CGO resolver._

**Reply by thanhdanh27600** [reply]
Hi, thank you for your support. I'm using the prebuilt Linux arm64.

I tried rebuilding the repository with CGO using the following command

`% GOOS=linux GOARCH=arm64 CGO_ENABLED=1 go build` 

However, I encountered the following errors:

```
linux_syscall.c:67:13: error: call to undeclared function 'setresgid'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
linux_syscall.c:67:13: note: did you mean 'setregid'?
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/unistd.h:593:6: note: 'setregid' declared here
linux_syscall.c:73:13: error: call to undeclared function 'setresuid'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
linux_syscall.c:73:13: note: did you mean 'setreuid'?
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/unistd.h:595:6: note: 'setreuid' declared here
```

I also tried:

`GOOS=android GOARCH=arm64 CGO_ENABLED=1 go build`

But then:

`gcc_android.c:6:10: fatal error: 'android/log.h' file not found`

If I omit the `GOOS` parameter, the build is successful, but when I try to run it on Termux (which is Linux-based), I get the following error:

`bash: ./base: cannot execute binary file: Exec format error`

I'm new to Go. I apologize for any confusion caused.

**Reply by ganigeorgiev** [reply]
As mentioned earlier, when `CGO_ENABLED=1` is used cross compilation may not work and you may have to run the build command on the device it self.

Another option you can try is to run the build command on your local machine with `CGO_ENABLED=0` as it seems that the CGO DNS resolver will be forced on Android even when CGO is not enabled, aka. try on your local machine disabling it and then upload the binary:

```js
GOOS=android GOARCH=arm64 CGO_ENABLED=0 go build
```
But I'm not sure if the above will work as the pure GO SQLite driver doesn't officially support Android and your best bet will be to build it with `CGO_ENABLED=1` on the Android device itself.

**Reply by ganigeorgiev** [reply]
If the prebuilt linux arm64 build runs fine for you (excluding the dns error), another option if you haven't tried it yet is to set the `GODEBUG=netdns=cgo` env variable:

```sh
GODEBUG=netdns=cgo ./pocketbase serve
```

**Reply by thanhdanh27600** [reply]
Thank @ganigeorgiev. Rebuild on local machine works!

```shell
$ pkg install golang
$ go version 
# go1.21.5 android/arm64
$ cd examples/base
$ GOOS=android GOARCH=arm64 CGO_ENABLED=1 go build
$ ./base serve
```

**Comment by VittorioParagallo** [comment]
Just in case anyone needs, i managed to run, on non rooted android, termux
serving 2 instances of pocketbase and nginx as reverse proxy. Certbot for
the https and goaccess to monitor. On the home router i set Duckdns and Nat
only the Port 443->2443 because on non rooted device i couldn't listen with
nginx on low ports. I used so https only till nginx (pocketbases were on
the sale host, so communicating on the sale host with nginx). Didn't try
with heavvy loads.. but It was working pretty good.
Just this morning that mobile device fell and broke :( so can't test It
more.

Il lun 25 dic 2023, 10:08 Phạm Thanh Danh ***@***.***> ha
scritto:

**Comment by jkdoshi** [comment]
Okay, I got this to work **_using the downloaded binary (without having to build my own)_**. Phew!

To make Golang runtime use Termux's resolv.conf, I did the following:

```
$ pkg install proot resolv-conf
$ proot -b $PREFIX/etc/resolv.conf:/etc/resolv.conf ./pocketbase ...
```
_Credit: https://stackoverflow.com/a/63915298_

That solved the DNS resolution problem, but only to reveal the next issue -- TLS/SSL cert trust-chains don't work!

```
$ proot -b $PREFIX/etc/resolv.conf:/etc/resolv.conf ./pocketbase update
Fetching release information...
Error: Get "https://api.github.com/repos/pocketbase/pocketbase/releases/latest": tls: failed to verify certificate: x509: certificate signed by unknown authority
````

So to fix that, I thought I'd just do the same for `/etc/tls` ... (but that didn't fix it) ...
```
$ proot -b $PREFIX/etc/resolv.conf:/etc/resolv.conf ./pocketbase update
Fetching release information...
Error: Get "https://api.github.com/repos/pocketbase/pocketbase/releases/latest": tls: failed to verify certificate: x509: certificate signed by unknown authority
````
Turns out binding to `/etc/ssl` instead of `/etc/tls` does the trick (It seems that's where Golang runtime is looking) ...
```
$ proot -b $PREFIX/etc/resolv.conf:/etc/resolv.conf -b $PREFIX/etc/tls:/etc/ssl ./pocketbase update
Fetching release information...
You already have the latest version 0.26.6.
```

So, in summary - Just use `chroot`/`proot` to bind/mount `/etc/resolve.conf` and `/etc/ssl` with the following prefix ...
```
$ proot -b $PREFIX/etc/resolv.conf:/etc/resolv.conf -b $PREFIX/etc/tls:/etc/ssl pocketbase ...
```
... and it works. It probably works for many other Golang programs too in Termux.

Hope it helps someone.
